File: matlab/core/parameters/CameraIntrinsics.m
classdef CameraIntrinsics
    %CAMERA_INTRINSICS basic parameters of camera
    
    properties
        camera_width    % width of image  [pixels]
        camera_height   % height of image [pixels]
        principal_point % [u,v]' center of image [pixels]
        focal_length    % [1/pixels or meter/pixels]
        baseline % NOT REALLY USED

        % optional properties
        % TODO: use these values in load_image()
        intensity_img_depth, depth_img_depth % for reading raw images and converting them to doubles in range [0, 1]
        near_clipping, far_clipping % for mapping raw depth values to 'real' values
    end
    
    methods
        function obj = CameraIntrinsics(width, height, focal)
            assert(width > 0);
            assert(height > 0);
            assert(focal > 0);
            
            assert(all(size(width)  == [1,1]));
            assert(all(size(height) == [1,1]));
            assert(all(size(focal)  == [1,1]));
            
            obj.camera_width    = width;
            obj.camera_height   = height;
            obj.focal_length    = focal;
            
            % [1,1] is at center of first pixel at the top left
            % [W,H] is at center of last pixel at the bottom right
            % i.e. integer coordinates directly corespond to pixel values
            %      (this is the same grid as interp2() uses)
            obj.principal_point = [width/2+0.5 height/2+0.5]';

            % default values for Blender
            obj.intensity_img_depth = 255;
            obj.depth_img_depth     = 255;
            obj.near_clipping       = 0.1;
            obj.far_clipping        = 100;
        end
    end

    methods(Static)
        function obj = loadFromCSV(input_path)
            parameters = csvread(fullfile(input_path, 'camera_intrinsics.csv'), 1, 0);

            % focal length, focal length mm, image width, image height, sensor width mm, sensor height mm, clip_start, clip_end

            obj = CameraIntrinsics(parameters(3), parameters(4), parameters(1));
            obj.near_clipping = parameters(7);
            obj.far_clipping  = parameters(8);
            obj.intensity_img_depth = parameters(9);
            obj.depth_img_depth     = parameters(10);
        end
    end
    
end

File: cpp/core/camera_intrinsics.cpp
#include "camera_intrinsics.h"

/*
 * example CSV file:
 *
 * focal length,focal length mm,image width,image height,sensor width mm,sensor height mm,clip start,clip end,color depth,depth depth
 * 280.0,35.0,256.0,128.0,32.0,18.0,0.10000000149011612,100.0,255,255
 *
 */

using namespace std;
using namespace boost;

///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
///////////////////////////////////////////////////////////////////////////////
CameraIntrinsics::CameraIntrinsics()
  : camera_width(0), camera_height(0),
    principal_point_x(0), principal_point_y(0),
    focal_length(0), baseline(0), is_disparity(false),
    near_clipping(0), far_clipping(0)
{
}
///////////////////////////////////////////////////////////////////////////////
CameraIntrinsics::CameraIntrinsics(unsigned int W, unsigned int H, float focal, float near, float far)
  : camera_width(W), camera_height(H),
    focal_length(focal), baseline(0), is_disparity(false),
    near_clipping(near), far_clipping(far)
{
    this->updatePrincipalPoint();
}
///////////////////////////////////////////////////////////////////////////////
CameraIntrinsics::CameraIntrinsics(const Eigen::Vector2f& camera_size, const Eigen::Vector2f& principal_point, float focal_length, float baseline)
  : camera_width(camera_size.x()), camera_height(camera_size.y()),
    principal_point_x(principal_point.x()), principal_point_y(principal_point.y()),
    focal_length(focal_length), baseline(baseline), is_disparity(true),
    near_clipping(0), far_clipping(0)
{
}
///////////////////////////////////////////////////////////////////////////////
void CameraIntrinsics::loadFromCSV(const std::string& filename)
{
    // read file
    ///////////////////////////////////

    ifstream inputfile(filename.c_str(), ifstream::in);

    if (!inputfile.is_open()) {
        // TODO: throw error 404
        cerr << "could not open CSV file '" << filename << "'" << endl;
        return;
    }

    // get second line with data
    string line;
    getline(inputfile, line);
    getline(inputfile, line);
    inputfile.close();

    assert(!line.empty());

    // parse data
    ///////////////////////////////////

    tokenizer< escaped_list_separator<char> > tok(line);

    tokenizer<escaped_list_separator<char> >::iterator it = tok.begin();

    this->focal_length  = lexical_cast<float>(trim(*it++));
    ++it; // skip focal length in mm
    this->camera_width  = lexical_cast<float>(trim(*it++)); // actually an uint, but Matlab/Blender writes this as float
    this->camera_height = lexical_cast<float>(trim(*it++));
    ++it; // skip sensor height
    ++it; // skip sensor width
    this->near_clipping = lexical_cast<float>(trim(*it++));
    this->far_clipping  = lexical_cast<float>(trim(*it++));
    // skip rest of csv

    // calculate other parameters
    ///////////////////////////////////

    updatePrincipalPoint();
}
///////////////////////////////////////////////////////////////////////////////
void CameraIntrinsics::updatePrincipalPoint()
{
    // [0, 0]     is at the center of the first pixel at the top left
    // [W-1, H-1] is at the center of the last pixel at the bottom right
    // i.e. integer coordinates directly corespond to pixel values
    this->principal_point_x = this->camera_width  / 2.0f - 0.5;
    this->principal_point_y = this->camera_height / 2.0f- 0.5;
}
///////////////////////////////////////////////////////////////////////////////
void CameraIntrinsics::downsample2()
{
    this->camera_width  /= 2;
    this->camera_height /= 2;
    this->focal_length  /= 2;
    this->updatePrincipalPoint();
}
///////////////////////////////////////////////////////////////////////////////
std::ostream& operator <<(std::ostream &output, const CameraIntrinsics &intrinsics)
{
    cout << "focal length = " << intrinsics.focal_length << endl;
    cout << "width = " << intrinsics.camera_width << "  height = " << intrinsics.camera_height << endl;
    cout << "principal point = " << intrinsics.principal_point_x << " " << intrinsics.principal_point_y << endl;
    cout << "clipping = " << intrinsics.near_clipping << " to " << intrinsics.far_clipping << endl;

    return output;
}
///////////////////////////////////////////////////////////////////////////////
